lib/active_fedora/associations/directly_contains_one_association.rb

```diff
@@ -0,0 +1,118 @@
+module ActiveFedora
+  module Associations
+    # Filters a DirectContainer relationship, returning the first item that matches the given :type
+    class DirectlyContainsOneAssociation < SingularAssociation #:nodoc:
+
+      # Finds objects contained by the container predicate (either the configured has_member_relation or ldp:contains)
+      # TODO: Refactor this to use solr (for efficiency) instead of parsing the RDF graph.  Requires indexing ActiveFedora::File objects into solr, including their RDF.type and, if possible, the id of their container
+      def find_target
+        # filtered_objects = container_association_proxy.to_a.select { |o| o.metadata_node.type.include?(options[:type]) }
+        query_node = if container_predicate = options[:has_member_relation]
+                       owner
+                     else
+                       container_predicate = ::RDF::Vocab::LDP.contains
+                       container_association.container  # Use the :through association's container
+                     end
+
+        contained_uris = query_node.resource.query(predicate: container_predicate).map { |r| r.object.to_s }
+        contained_objects = contained_uris.map { |object_uri| klass.find(klass.uri_to_id(object_uri)) }
+        filtered_objects = contained_objects.select {|o| o.metadata_node.type.include?(options[:type]) }
+
+        return filtered_objects.first
+      end
+
+      # # Build/Create record and ensure that it's added to the corresponding Container
+      def new_record(method, attributes={})
+        # This works, using the container's builder, but then you're stuck with the container's class_name
+        # record = @owner.send(container_reflection.name).send(method, attributes) do |record|
+        #   add_type_to_record(record, options[:type])
+        # end
+
+        # Build the record using _this_ association's build method (so setting things like class_name on this association will have an effect)
+        record = @reflection.send("#{method}_association", attributes)
```

Comment:
Do we need to override this method? It seem that the super will call replace which should add_to_container. See comment below (line 60) regarding moving `add_type_to_record` to `replace` too.